modules/apps/callbacks/callbacks_test.go

```diff
@@ -0,0 +1,263 @@
+	s.path.EndpointB.ChannelConfig.PortID = icatypes.HostPortID
+	s.path.EndpointA.ChannelConfig.Version = icaVersion
+	s.path.EndpointB.ChannelConfig.Version = icaVersion
+
+	s.RegisterInterchainAccount(icaOwner)
+	// open chan init must be skipped. So we cannot use .CreateChannels()
+	err = s.path.EndpointB.ChanOpenTry()
+	s.Require().NoError(err)
+	err = s.path.EndpointA.ChanOpenAck()
+	s.Require().NoError(err)
+	err = s.path.EndpointB.ChanOpenConfirm()
+	s.Require().NoError(err)
+
+	interchainAccountAddr, found := s.chainB.GetSimApp().ICAHostKeeper.GetInterchainAccountAddress(s.chainB.GetContext(), s.path.EndpointA.ConnectionID, s.path.EndpointA.ChannelConfig.PortID)
+	s.Require().True(found)
+
+	// fund the interchain account on chainB
+	msgBankSend := &banktypes.MsgSend{
+		FromAddress: s.chainB.SenderAccount.GetAddress().String(),
+		ToAddress:   interchainAccountAddr,
+		Amount:      sdk.NewCoins(sdk.NewCoin(sdk.DefaultBondDenom, sdkmath.NewInt(100000))),
+	}
+	res, err := s.chainB.SendMsgs(msgBankSend)
+	s.Require().NotEmpty(res)
+	s.Require().NoError(err)
+
+	return interchainAccountAddr
+}
+
+// RegisterInterchainAccount submits a MsgRegisterInterchainAccount and updates the controller endpoint with the
+// channel created.
+func (s *CallbacksTestSuite) RegisterInterchainAccount(owner string) {
+	msgRegister := icacontrollertypes.NewMsgRegisterInterchainAccount(s.path.EndpointA.ConnectionID, owner, s.path.EndpointA.ChannelConfig.Version)
+
+	res, err := s.chainA.SendMsgs(msgRegister)
+	s.Require().NotEmpty(res)
+	s.Require().NoError(err)
+
+	channelID, err := ibctesting.ParseChannelIDFromEvents(res.Events)
+	s.Require().NoError(err)
+
+	s.path.EndpointA.ChannelID = channelID
+}
+
+// AssertHasExecutedExpectedCallback checks the stateful entries and counters based on callbacktype.
+// It assumes that the source chain is chainA and the destination chain is chainB.
+func (s *CallbacksTestSuite) AssertHasExecutedExpectedCallback(callbackType types.CallbackType, expSuccess bool) {
+	var expStatefulEntries uint8
+	if expSuccess {
+		// if the callback is expected to be successful,
+		// we expect at least one state entry
+		expStatefulEntries = 1
+	}
+
+	sourceStatefulCounter := s.chainA.GetSimApp().MockContractKeeper.GetStateEntryCounter(s.chainA.GetContext())
+	destStatefulCounter := s.chainB.GetSimApp().MockContractKeeper.GetStateEntryCounter(s.chainB.GetContext())
+
+	switch callbackType {
+	case "none":
+		s.Require().Equal(uint8(0), sourceStatefulCounter)
+		s.Require().Equal(uint8(0), destStatefulCounter)
+
+	case types.CallbackTypeSendPacket:
+		s.Require().Equal(expStatefulEntries, sourceStatefulCounter, "unexpected stateful entry amount for source send packet callback")
+		s.Require().Equal(uint8(0), destStatefulCounter)
+
+	case types.CallbackTypeAcknowledgementPacket, types.CallbackTypeTimeoutPacket:
+		expStatefulEntries *= 2 // expect OnAcknowledgement/OnTimeout to be successful as well
```

Comment:
The `*2` is because `SendPacket` is ultimately called first, right? Maybe we should add that to the comment

```suggestion
		expStatefulEntries *= 2 // expect OnAcknowledgement/OnTimeout to be successful as well as the initial SendPacket
```

Reply:
done